src/serialization/serialization.rs

```diff
@@ -0,0 +1,412 @@
+        // TODO benchmark minimizing read calls by reading into a fixed-size header buffer
+
+        let cookie = reader.read_u32::<BigEndian>()?;
+
+        if cookie != V2_COOKIE {
+            return Err(DeserializeError::InvalidCookie);
+        }
+
+        let payload_len = reader.read_u32::<BigEndian>()?.to_usize()
+            .ok_or(DeserializeError::UsizeTypeTooSmall)?;
+        let normalizing_offset = reader.read_u32::<BigEndian>()?;
+        if normalizing_offset != 0 {
+            return Err(DeserializeError::UnsupportedFeature);
+        }
+        let num_digits = reader.read_u32::<BigEndian>()?.to_u8()
+            .ok_or(DeserializeError::InvalidParameters)?;
+        let low = reader.read_u64::<BigEndian>()?;
+        let high = reader.read_u64::<BigEndian>()?;
+        let int_double_ratio = reader.read_f64::<BigEndian>()?;
+        if int_double_ratio != 1.0 {
+            return Err(DeserializeError::UnsupportedFeature);
+        }
+
+        let mut h = Histogram::new_with_bounds(low, high, num_digits)
+            .map_err(|_| DeserializeError::InvalidParameters)?;
+
+        if payload_len > self.payload_buf.len() {
+            self.payload_buf.resize(payload_len, 0);
+        }
+
+        let mut payload_slice = &mut self.payload_buf[0..payload_len];
+        reader.read_exact(&mut payload_slice)?;
+
+        let mut cursor = Cursor::new(&payload_slice);
+        let mut dest_index: usize = 0;
+        while cursor.position() < payload_slice.len() as u64 {
+            let num = zig_zag_decode(varint_read(&mut cursor)?);
+
+            if num < 0 {
+                let zero_count = (-num).to_usize()
+                    .ok_or(DeserializeError::UsizeTypeTooSmall)?;
+                // skip the zeros
+                dest_index = dest_index.checked_add(zero_count)
+                    .ok_or(DeserializeError::UsizeTypeTooSmall)?;
+                continue;
+            } else {
+                let count: T = T::from_i64(num)
+                    .ok_or(DeserializeError::UnsuitableCounterType)?;
+
+                h.set_count_at_index(dest_index, count)
+                    .map_err(|_| DeserializeError::EncodedArrayTooLong)?;
+
+                dest_index = dest_index.checked_add(1)
+                    .ok_or(DeserializeError::UsizeTypeTooSmall)?;
+            }
+        }
+
+        // TODO destat is expensive; should accumulate the necessary state while deserializing
+        // dest_index is one past the last written index, and is therefore the length to scan
+        h.restat(dest_index);
+
+        Ok(h)
+    }
+}
+
+/// Encode counts array into slice.
+/// The slice must be at least 9 * the number of counts that will be encoded.
+fn encode_counts<T: Counter>(h: &Histogram<T>, buf: &mut [u8]) -> Result<usize, SerializeError> {
+    let index_limit = h.index_for(h.max()) + 1;
+    let mut index = 0;
+    let mut bytes_written = 0;
+
+    assert!(index_limit <= h.counts.len());
+
+    while index < index_limit {
```

Comment:
Actually, as written it's not quite right. You need to `.peek()`, and then call `.next()` if you actually end up taking the value, otherwise the iterator won't progress.